runtime/lua/vim/keymap.lua

```diff
@@ -0,0 +1,138 @@
+local keymap = {}
+
+--- Add a new |mapping|.
+--- Examples:
+--- <pre>
+---   -- Can add mapping to Lua functions
+---   vim.keymap.set('n', 'lhs', function() print("real lua function") end)
+---
+---   -- Can use it to map multiple modes
+---   vim.keymap.set({'n', 'v'}, '<leader>lr', vim.lsp.buf.references, { buffer=true })
+---
+---   -- Can add mapping for specific buffer
+---   vim.keymap.set('n', '<leader>w', "<cmd>w<cr>", { silent = true, buffer = 5 })
+---
+---   -- Expr mappings
+---   vim.keymap.set('i', '<Tab>', function()
+---     return vim.fn.pumvisible() == 1 and "<C-n>" or "<Tab>"
+---   end, { expr = true })
+---   -- <Plug> mappings
+---   vim.keymap.set('n', '[%%', '<Plug>(MatchitNormalMultiBackward)')
+--- </pre>
+---
+--- Note that in a mapping like:
+--- <pre>
+---    vim.keymap.set('n', 'asdf', require('jkl').my_fun)
+--- </pre>
+---
+--- the require('jkl') gets evaluated during this call in order to access the function. If you want to
+--- avoid this cost at startup you can wrap it in a function, for example:
+--- <pre>
+---    vim.keymap.set('n', 'asdf', function() return require('jkl').my_fun() end)
+--- </pre>
+---
+---@param mode string|table   Same mode short names as |nvim_set_keymap()|.
+---                            Can also be list of modes to create mapping on multiple modes.
+---@param lhs string          Left-hand side |{lhs}| of the mapping.
+---@param rhs string|function  Right-hand side |{rhs}| of the mapping. Can also be a Lua function.
+--
+---@param opts table A table of |:map-arguments| such as "silent". In addition to the options
+---                  listed in |nvim_set_keymap()|, this table also accepts the following keys:
+---                  - replace_termcodes: (boolean, default true) When both this and expr is "true",
+---                  |nvim_replace_termcodes()| is applied to the result of Lua expr maps.
+---                  - remap: (boolean) Make the mapping recursive. This is the
+---                  inverse of the "noremap" option from |nvim_set_keymap()|.
+---                  Default `true` if `lhs` is a string starting with `<plug>` (case-insensitive), `false` otherwise.
+---@see |nvim_set_keymap()|
+function keymap.set(mode, lhs, rhs, opts)
+  vim.validate {
+    mode = {mode, {'s', 't'}},
+    lhs = {lhs, 's'},
+    rhs = {rhs, {'s', 'f'}},
+    opts = {opts, 't', true}
+  }
+
+  opts = opts or {}
+  local buffer = false
+  local is_rhs_lua = type(rhs) == "function"
+  mode = type(mode) == 'string' and {mode} or mode
+
+  if opts.buffer ~= nil then
+    buffer = opts.buffer == true and 0 or opts.buffer
+    opts.buffer = nil
+  end
+
+  if is_rhs_lua and opts.expr and opts.replace_termcodes ~= false then
+    local user_rhs = rhs
+    rhs = function ()
+      return vim.api.nvim_replace_termcodes(user_rhs(), true, true, true)
```

Comment:
Hi. May I recommend `vim.api.nvim_replace_termcodes(vim.api.nvim_eval(user_rhs()), true, true, true)` so mappings like:
```lua
vim.keymap.set('<tab>', function()
  return "luasnip#expand_or_jumpable() ? '<Plug>luasnip-expand-or-jump' : '<tab>'"
end, { expr = true })
```
would work.

Reply:
This specific snippet is not a compelling example since it can be completely translated to a Lua (expr) mapping.

Reply:
I'd say no because then lua functions directly returning keycodes will not work .

Reply:
or even 
```lua
vim.keymap.set('i', '<tab>', function()
  if require"luasnip".expand_or_jumpable() then
    require'luasnip'.expand_or_jump()
  else
    return "<tab>"
  end
end, { expr = true })
```

Reply:
(which I assume LuaSnip will eventually implement themselves)

Reply:
> No because then lua functions directly returning keycodes will not work .
> 
> Also I can't see why you'd be wraping that in `function() end` ? Why not just use
> 
> ```lua
> vim.keymap.set('i', '<tab>', "luasnip#expand_or_jumpable() ? '<Plug>luasnip-expand-or-jump' : '<tab>'", { expr = true })
> ```

True, it wouldn't be reasonable to wrap such a mapping in a lua function.

Reply:
> (which I assume LuaSnip will eventually implement themselves)

Yes, I use a similar mapping.

Reply:
> or even
> 
> ```lua
> vim.keymap.set('i', '<tab>', function()
>   if require"luasnip".expand_or_jumpable() then
>     require'luasnip'.expand_or_jump()
>   else
>     return "<tab>"
>   end
> end, { expr = true })
> ```

Yes, I use a similar mapping.

Reply:
> or even
> 
> ```lua
> vim.keymap.set('i', '<tab>', function()
>   if require"luasnip".expand_or_jumpable() then
>     require'luasnip'.expand_or_jump()
>   else
>     return "<tab>"
>   end
> end, { expr = true })
> ```

Hopefully you will forgive the off-topic question, but does this mapping work for you?
I've been trying unsuccessfully so far to port my luasnip keybinds but no luck so far.. the above example produces a `Failed to save undo information` error.